src/test/mir-opt/simplify-arm-identity/rustc.main.SimplifyArmIdentity.diff

```diff
@@ -33,15 +33,14 @@
       }
   
       bb1: {
-          ((_2 as Foo).0: u8) = const 0u8; // scope 1 at $DIR/simplify-arm-identity.rs:20:21: 20:32
+          _2 = const Dst::Foo(0u8);        // bb1[0]: scope 1 at $DIR/simplify-arm-identity.rs:20:21: 20:32
```

Comment:
:rofl: it took me a bit to figure this one out. Since we're now not unilaterally marking `_2 = Dst::Foo(u8);` as unpromotable, we're changing it to `_2 = const Dst::Foo(u8);`, making deaggregation later fail to deaggregate. That's actually pretty neat.

cc @RalfJung this is a cool effect of having deaggregation after const prop

Reply:
For cg_clif this would mean having to `memcpy` from the `.rodata` section to the stack.

Reply:
Is that for all non-primitive scalars? Or for all constants in general?

Reply:
For all constants that are not stored in ssa variables, aka everything that is not a bool, int, char, float, function pointer or thin reference/pointer.

https://github.com/bjorn3/rustc_codegen_cranelift/blob/456c14f9e8c4e42fd97730c33a39833c124de8ad/src/constant.rs#L113
https://github.com/bjorn3/rustc_codegen_cranelift/blob/456c14f9e8c4e42fd97730c33a39833c124de8ad/src/common.rs#L37

Reply:
hmm... so you'd prefer for simple constants that they are generated by code instead of `memcpy`d over? We could probably build a MIR pass like that that cg_clif could turn on.

Reply:
> so you'd prefer for simple constants that they are generated by code instead of `memcpy`d over?

Yes. I think I can do this in cg_clif itself now that I think about it though. I opened https://github.com/bjorn3/rustc_codegen_cranelift/issues/1041 for this.

Reply:
@bjorn3 It sounds like your concern can be resolved in cg_clif. Will it cause any issue for you if we go ahead and merge this?

Reply:
It is fine to merge this. It will just cause a bit of a performance hit for now.

Reply:
> this is a cool effect of having deaggregation after const prop

Why would ConstProp introduce new aggregated assignments if previously everything was deaggregated?

Reply:
It won't. It turns `Rvalue::Aggregate` into `Rvalue::Use(Operand::Const)`, so if we still have aggregates at const prop, we'll turn those into consts that we can, while after deaggregation we can't figure this out and keep all the individual assignments

Reply:
That sounds like the current ConstProp was just geared towards the kind of MIR it currently sees, but not like a fundamental limitation?

Reply:
Unless we add a mir statement for "this type is now fully initialized", I don't see how const prop can ever know for sure that it can now generate a constant that initializes the entire local and erase all the individual steps.

I guess we could run validation and check if that succeeds, but that seems awefully ad-hoc.

Reply:
Oh, so we actually *want* these assignments to remain aggregated? (Assigning a large const seems very similar to an aggregated assignment to me.) Interesting.

Reply:
I'll open an issue about this. I think we do, but I don't think we ever talked about it.

Reply:
https://github.com/rust-lang/rust/issues/73203